src/TestInfrastructure/Helpers/TestDomainWrapper.cs

```diff
@@ -0,0 +1,159 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+/* Assembly resolution:
+ * We want assembly resolution in the test app domain to work the same way it
+ * does in the main/calling app domain. To do this, we set up assembly resolver
+ * for the test app domain to handle locating any assemblies the runtime fails
+ * to find. The resolver method makes a call to code running in the main app
+ * domain, and we try to locate the assembly in the main app domain.
+ */
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
+{
+    /// <summary>
+    /// Helper class that provides the plumbing to run a test in a separate app domain
+    /// </summary>
+    /// <typeparam name="T">The type of the remote object to be created the new app domain</typeparam>
+    /// <remarks>The new AppDomain will be created when the test wrapper instance is created,
+    /// and unloaded when the test wrapper is disposed.</remarks>
+    public sealed class TestDomainWrapper<T> : MarshalByRefObject, IDisposable
+        where T : MarshalByRefObject
+    {
+        private const string TestDomainName = "sonarlint remote test domain";
+        private const string AppDomainDataKey = "test domain wrapper";
+
+        public TestDomainWrapper()
+        {
+            CreateTestAppDomain();
+        }
+
+        public AppDomain TestAppDomain { get; private set; }
+
+        /// <summary>
+        /// Any method calls made on the remote object will be executed in the remote
+        /// app domain
+        /// </summary>
+        public T RemoteObject { get; private set; }
+
+        private void CreateTestAppDomain()
+        {
+            AppDomainSetup domainSetup = new AppDomainSetup()
+            {
+                ApplicationBase = AppDomain.CurrentDomain.BaseDirectory,
+            };
+
+            TestAppDomain = AppDomain.CreateDomain(TestDomainName, null, domainSetup);
+
+            TestAppDomain.SetData(AppDomainDataKey, this);
+
+            TestAppDomain.AssemblyResolve += OnAssemblyResolve;
+
+            RemoteObject = (T)TestAppDomain.CreateInstanceAndUnwrap(typeof(T).Assembly.FullName, typeof(T).FullName);
+        }
+
+        private static Assembly OnAssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            CheckIsInTestDomain();
+
+            TestDomainWrapper<T> resolver = AppDomain.CurrentDomain.GetData(AppDomainDataKey) as TestDomainWrapper<T>;
+            Assert.IsNotNull(resolver, "Test setup error: failed to obtain the remote domain wrapper");
```

Comment:
Feel free to ignore but I would use _FluentAssertions_ here.

Reply:
The message formatting used by _FluentAssertions_ doesn't make sense in this case. I've changed them to throw _Exception_.
FYI this is a habit from MS days: when triaging test failures, it made a difference whether the failure was in test code or in product code, so we wanted it to be obvious from the failure message which it was.